File: sdk-ruby/lib/daytona/secret_service.rb
# Copyright Daytona Platforms Inc.
# SPDX-License-Identifier: Apache-2.0

# frozen_string_literal: true

module Daytona
  class SecretService
    include Instrumentation

    # Service for managing organization-scoped Daytona Secrets. Can be used to list, get, create,
    # update and delete Secrets.
    #
    # A Secret stores a plaintext +value+ that is never returned by the API. When a Secret is
    # referenced while creating a Sandbox, the corresponding env var holds an opaque +placeholder+
    # that is resolved to the real value only for the Secret's allowed +hosts+.
    #
    # @param secret_api [DaytonaApiClient::SecretApi]
    # @param otel_state [Daytona::OtelState, nil]
    def initialize(secret_api, otel_state: nil)
      @secret_api = secret_api
      @otel_state = otel_state
    end

    # Create a new Secret.
    #
    # @param name [String] Name of the Secret. Must match +^[a-zA-Z_][a-zA-Z0-9_-]*$+ and be unique
    #   within the organization (a duplicate name raises a 409 error).
    # @param value [String] Plaintext value of the Secret. Write-only; never returned by the API.
    # @param description [String, nil] Optional description of the Secret.
    # @param hosts [Array<String>, nil] Allowed hosts this Secret may be sent to. Accepts exact
    #   hostnames and +*.+ wildcards (no ports).
    # @return [Daytona::Secret]
    def create(name, value, description: nil, hosts: nil)
      Secret.new(secret_api.create_secret(
                   DaytonaApiClient::CreateSecret.new(name:, value:, description:, hosts:)
                 ))
    end

    # Delete a Secret.
    #
    # @param secret_id [String]
    # @return [void]
    # @raise [DaytonaApiClient::ApiError] If no Secret with the given ID exists (404).
    def delete(secret_id) = secret_api.delete_secret(secret_id)

    # Get a Secret by ID.
    #
    # @param secret_id [String]
    # @return [Daytona::Secret]
    # @raise [DaytonaApiClient::ApiError] If no Secret with the given ID exists (404).
    def get(secret_id) = Secret.new(secret_api.get_secret(secret_id))

    # List Secrets with cursor-based pagination.
    #
    # @param cursor [String, nil] Pagination cursor from a previous response.
    # @param limit [Integer, nil] Number of results per page (1-200, defaults to 100).
    # @param name [String, nil] Filter by partial name match.
    # @param sort [String, nil] Field to sort by: +name+, +createdAt+ or +updatedAt+
    #   (defaults to +createdAt+).
    # @param order [String, nil] Direction to sort by: +asc+ or +desc+ (defaults to +desc+).
    # @return [Daytona::ListSecretsResponse]
    # @raise [Daytona::Sdk::Error]
    #
    # @example
    #   daytona = Daytona::Daytona.new
    #   cursor = nil
    #   loop do
    #     page = daytona.secret.list(cursor:, limit: 100)
    #     page.items.each { |secret| puts secret.name }
    #     cursor = page.next_cursor
    #     break if cursor.nil?
    #   end
    def list(cursor: nil, limit: nil, name: nil, sort: nil, order: nil)
      raise Sdk::Error, 'limit must be positive integer' if limit && limit < 1

      response = secret_api.list_secrets_paginated(cursor:, limit:, name:, sort:, order:)
      ListSecretsResponse.new(
        items: response.items.map { |secret| Secret.new(secret) },
        total: response.total,
        next_cursor: response.next_cursor
      )
    end

    # Update a Secret.
    #
    # @param secret_id [String]
    # @param value [String, nil] New plaintext value. Write-only; never returned by the API.
    # @param description [String, nil] New description of the Secret.
    # @param hosts [Array<String>, nil] Allowed hosts this Secret may be sent to. Accepts exact
    #   hostnames and +*.+ wildcards (no ports).
    # @return [Daytona::Secret]
    # @raise [DaytonaApiClient::ApiError] If no Secret with the given ID exists (404).
    def update(secret_id, value: nil, description: nil, hosts: nil)
      Secret.new(secret_api.update_secret(
                   secret_id,
                   DaytonaApiClient::UpdateSecret.new(value:, description:, hosts:)
                 ))
    end

    instrument :create, :delete, :get, :list, :update, component: 'SecretService'

    private

    # @return [DaytonaApiClient::SecretApi]
    attr_reader :secret_api

    # @return [Daytona::OtelState, nil]
    attr_reader :otel_state
  end
end

File: sdk-java/src/main/java/io/daytona/sdk/SecretService.java
// Copyright Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

package io.daytona.sdk;

import io.daytona.api.client.api.SecretApi;
import io.daytona.api.client.model.CreateSecret;
import io.daytona.api.client.model.UpdateSecret;
import io.daytona.sdk.model.CreateSecretParams;
import io.daytona.sdk.model.ListSecretsQuery;
import io.daytona.sdk.model.ListSecretsResponse;
import io.daytona.sdk.model.Secret;
import io.daytona.sdk.model.UpdateSecretParams;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

/**
 * Service for managing organization-scoped Daytona Secrets.
 *
 * <p>Secrets can be created, listed, retrieved, updated, and deleted, and referenced when creating a
 * Sandbox via the {@code secrets} field on the create-sandbox parameters. The plaintext {@code value}
 * is write-only and is never returned by the API; the Sandbox only ever sees the Secret's opaque
 * placeholder, and the real value is substituted at the network egress layer for the Secret's allowed
 * hosts.
 */
public class SecretService {
    private final SecretApi secretApi;

    SecretService(SecretApi secretApi) {
        this.secretApi = secretApi;
    }

    /**
     * Creates a new Secret.
     *
     * @param params creation parameters; {@code name} must match {@code ^[a-zA-Z_][a-zA-Z0-9_-]*$} and
     *               be unique within the organization
     * @return created {@link Secret} (without the plaintext {@code value})
     * @throws io.daytona.sdk.exception.DaytonaConflictException if a Secret with the same name already exists
     * @throws io.daytona.sdk.exception.DaytonaException if creation fails
     */
    public Secret create(CreateSecretParams params) {
        CreateSecret req = new CreateSecret().name(params.getName()).value(params.getValue());
        if (params.getDescription() != null) {
            req.setDescription(params.getDescription());
        }
        if (params.getHosts() != null) {
            req.setHosts(params.getHosts());
        }
        io.daytona.api.client.model.Secret secret = ExceptionMapper.callMain(
                () -> secretApi.createSecret(req, null)
        );
        return toSecret(secret);
    }

    /**
     * Lists Secrets in the organization one page at a time, using default query parameters.
     *
     * @return page of Secrets; {@code nextCursor} is {@code null} when there are no more pages
     * @throws io.daytona.sdk.exception.DaytonaException if the API request fails
     */
    public ListSecretsResponse list() {
        return list(null);
    }

    /**
     * Lists Secrets in the organization one page at a time. Pass the {@code nextCursor} from a
     * previous response as the query {@code cursor} to fetch the next page.
     *
     * @param query optional filter, sort, and pagination parameters; may be {@code null}
     * @return page of Secrets; {@code nextCursor} is {@code null} when there are no more pages
     * @throws io.daytona.sdk.exception.DaytonaException if the API request fails
     */
    public ListSecretsResponse list(ListSecretsQuery query) {
        String cursor = query == null ? null : query.getCursor();
        BigDecimal limit = query == null || query.getLimit() == null ? null : BigDecimal.valueOf(query.getLimit());
        String name = query == null ? null : query.getName();
        String sort = query == null ? null : query.getSort();
        String order = query == null ? null : query.getOrder();
        io.daytona.api.client.model.ListSecretsResponse result = ExceptionMapper.callMain(
                () -> secretApi.listSecretsPaginated(null, cursor, limit, name, sort, order)
        );

        ListSecretsResponse output = new ListSecretsResponse();
        List<Secret> items = new ArrayList<Secret>();
        if (result != null && result.getItems() != null) {
            for (io.daytona.api.client.model.Secret secret : result.getItems()) {
                items.add(toSecret(secret));
            }
        }
        output.setItems(items);
        output.setTotal(result != null && result.getTotal() != null ? result.getTotal().intValue() : 0);
        output.setNextCursor(result == null ? null : result.getNextCursor());
        return output;
    }

    /**
     * Retrieves a Secret by ID.
     *
     * @param secretId Secret identifier
     * @return matching {@link Secret}
     * @throws io.daytona.sdk.exception.DaytonaNotFoundException if no Secret with the given ID exists
     * @throws io.daytona.sdk.exception.DaytonaException if the request fails
     */
    public Secret get(String secretId) {
        io.daytona.api.client.model.Secret secret = ExceptionMapper.callMain(() -> secretApi.getSecret(secretId, null));
        return toSecret(secret);
    }

    /**
     * Updates an existing Secret. Omitted ({@code null}) fields are left unchanged.
     *
     * @param secretId Secret identifier
     * @param params fields to update
     * @return updated {@link Secret}
     * @throws io.daytona.sdk.exception.DaytonaNotFoundException if no Secret with the given ID exists
     * @throws io.daytona.sdk.exception.DaytonaException if the request fails
     */
    public Secret update(String secretId, UpdateSecretParams params) {
        UpdateSecret req = new UpdateSecret();
        if (params.getValue() != null) {
            req.setValue(params.getValue());
        }
        if (params.getDescription() != null) {
            req.setDescription(params.getDescription());
        }
        if (params.getHosts() != null) {
            req.setHosts(params.getHosts());
        }
        io.daytona.api.client.model.Secret secret = ExceptionMapper.callMain(
                () -> secretApi.updateSecret(secretId, req, null)
        );
        return toSecret(secret);
    }

    /**
     * Deletes a Secret by ID.
     *
     * @param secretId Secret identifier
     * @throws io.daytona.sdk.exception.DaytonaNotFoundException if no Secret with the given ID exists
     * @throws io.daytona.sdk.exception.DaytonaException if deletion fails
     */
    public void delete(String secretId) {
        ExceptionMapper.runMain(() -> secretApi.deleteSecret(secretId, null));
    }

    private Secret toSecret(io.daytona.api.client.model.Secret source) {
        Secret secret = new Secret();
        if (source != null) {
            secret.setId(source.getId());
            secret.setName(source.getName());
            secret.setDescription(source.getDescription());
            secret.setPlaceholder(source.getPlaceholder());
            secret.setHosts(source.getHosts());
            secret.setCreatedAt(source.getCreatedAt() == null ? null : source.getCreatedAt().toString());
            secret.setUpdatedAt(source.getUpdatedAt() == null ? null : source.getUpdatedAt().toString());
        }
        return secret;
    }
}
